Jetpack Cloud Agency Dashboard: add issue license and add new site buttons to resting state

TheSteveK opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to # 1199980337313591-as-1204977192508984/f

## Proposed Changes

* We have been asked to add a new set of buttons that display in the "default" view of the dashboard
* The buttons are add new site and issue a license
* The add new site button goes to wordpress.com/jetpack/connect
* The issue license page goes to the licenses tab on the dashboard.
* On desktop, the buttons appear in the same location as the existing Add x Licenses button
* On mobile, the buttons appear in a new section above the sites list tab
* This is a new section, but some things had to be altered in the code to make it happen.
* A section had to be added to the mobile version
* The existing buttons on desktop were sticky and appeared conditionally.
* The header component that housed the buttons was configured to only show sticky content. I removed that restriction so it would show whatever content was called for.
* Added conditions in the render to display the new buttons when no license was selected, removed a variable to detect screen size and if licenses were selected as it seemed unnecessary based on the new way of doing things.

The new buttons in pictures:

![Screenshot 2023-07-25 at 9 20 04 AM](https://github.com/Automattic/wp-calypso/assets/12895386/bf9a544c-c10c-41e8-a33e-9ad8e9ae48dd)
![Screenshot 2023-07-25 at 9 19 21 AM](https://github.com/Automattic/wp-calypso/assets/12895386/85b1b9b6-80eb-45ff-9bf9-1533079d2404)


## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Your account must be an agency partner to see the dashboard
* Head over to the Jetpack Cloud Link (or check out this PR and run locally, I won't judge)
* You should automatically be redirected to /dashboard
* Observe the new buttons.  Click each to verify add site goes to wordpress.com/jetpack/connect and issue license goes to the license tab.
* Next to one of your sites, click the add button in one of the columns.  The original add x licenses button should now show in the same place as the new ones
* Do the same on the mobile view (or desktop if you started with mobile)
* Expand a site and click add next to one of the available features, the new buttons should go away and the add x licenses should appear at the bottom of the page

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

TheSteveK committed:
update CSS

TheSteveK committed:
Update CSS for large screen view

TheSteveK committed:
fix sticky issue add tracks events

TheSteveK committed:
remove commented out code

TheSteveK committed:
CSS changes and code cleanup

TheSteveK committed:
fix extra spaces in scss

TheSteveK committed:
make SiteContentHeader conditional on screen size

TheSteveK commented:
it looks like one of the tests failed because the new code is rendering something if the dashboard is in the empty state.  I'm going to push another commit to hide the buttons in a blank state

TheSteveK committed:
add empty state condition to new button component function

TheSteveK committed:
show ContentHeader hide content on mobile

yashwin commented:
@TheSteveK: I didn't review the whole PR. One suggestion: We could still keep the `Issue License` and `Add New Site` buttons on mobile views if that is easy to implement. We anyway don't utilize that space to show the buttons that appear when a license is selected.

yashwin commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/index.tsx:
```diff
@@ -191,6 +191,36 @@ export default function SitesOverview() {
+					href="/partner-portal/issue-license"
+					onClick={ () =>
+						dispatch(
+							recordTracksEvent( 'calypso_jetpack_agency_dashboard_issue_license_button_click', {} )
```
I don't think an empty object is needed here

```suggestion
							recordTracksEvent( 'calypso_jetpack_agency_dashboard_issue_license_button_click' )
```

yashwin commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/index.tsx:
```diff
@@ -191,6 +191,36 @@ export default function SitesOverview() {
+					href="https://wordpress.com/jetpack/connect"
+					onClick={ () =>
+						dispatch(
+							recordTracksEvent( 'calypso_jetpack_agency_dashboard_add_site_button_click', {} )
```
```suggestion
							recordTracksEvent( 'calypso_jetpack_agency_dashboard_add_site_button_click' )
```

TheSteveK replied:
ok, makes sense :)

yashwin commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/index.tsx:
```diff
@@ -191,6 +191,36 @@ export default function SitesOverview() {
+				</Button>
+				<Button
+					primary
+					className="sites-overview__add-site-button"
```
I think className is not needed here.

TheSteveK replied:
Ah right, because I'm using primary to target the button.

yashwin reviewed (commented):
It works as expected. 

One suggestion: We could still keep the `Issue License` and `Add New Site` buttons on mobile views if that is easy to implement. We anyway don't utilize that space to show the buttons that appear when a license is selected. WDYT?

I have also left a few comments.

yashwin commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/index.tsx:
```diff
@@ -229,25 +259,30 @@ export default function SitesOverview() {
 							pageTitle={ pageTitle }
-							showStickyContent={ !! showIssueLicenseButtonsLargeScreen }
+							// Only renderIssueLicenseButton should be sticky.
+							showStickyContent={ selectedLicensesCount > 0 }
```
I think we need `isWithinBreakpoint( '>960px' )` check here. WDYT?

TheSteveK replied:
Yes indeed, that was causing double buttons some times.

vitozev reviewed (changes requested):
It looks like there's still some work required to improve the responsiveness of the UI.

See the video below:


https://github.com/Automattic/wp-calypso/assets/1749918/77f6eccd-74be-4d88-a9ee-e8385dcba6d6

There's a spectrum of devices out there, so it's the responsiveness that matters when we work on the UI/UX.

TheSteveK committed:
fix sticky logic, small CSS update

TheSteveK commented:
> One suggestion: We could still keep the `Issue License` and `Add New Site` buttons on mobile views if that is easy to implement. We anyway don't utilize that space to show the buttons that appear when a license is selected. WDYT?

Good point, but it's a bit unclear from the previous discussions what the behavior should be.  I've asked Ilona to clarify in the original task and we'll go with whatever she says :)

TheSteveK committed:
Change query for large screen to a variable to simplify things

TheSteveK committed:
Add comment about mobile render

TheSteveK commented:
Cleaned up some more code.  Improved responsive layout with the last few PRs.  Also, I've implemented @yashwin suggestion about rendering the mobile buttons even if the add x licenses is active.  I haven't heard back on my design question, so I'll wait a bit before requesting a re-review.  I also see the teamcity builds are failing so I'll investigate that.  Maybe need to rebase.

TheSteveK committed:
Fix typescript error

TheSteveK committed:
Slight tweak to sticky code

TheSteveK commented:
Tests are good, Design is approved.  @yashwin I've implemented all the suggestions.  Let's try this again :)

vitozev commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/index.tsx:
```diff
@@ -242,13 +270,24 @@ export default function SitesOverview() {
 						<SiteContentHeader
-							content={ renderIssueLicenseButton() }
+							content={
+								// don't render content on mobile, we don't need it
```
Nitpick: When you read this comment, it looks like we don't display these buttons on mobile, which is incorrect.

TheSteveK replied:
Alright, I've changed the wording, hopefully it's a little more clear.

TheSteveK committed:
clarify comment on mobile screen render

mattgawarecki commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/index.tsx:
```diff
@@ -191,6 +191,35 @@ export default function SitesOverview() {
 		} );
 	}, [ selectedLicensesSiteId, selectedLicenses ] );
 
+	const renderAddSiteIssueLicenseButtons = () => {
+		return (
+			<div className="sites-overview__add-site-issue-license-buttons">
+				<Button
+					className="sites-overview__issue-license-button"
+					href="/partner-portal/issue-license"
+					onClick={ () =>
+						dispatch(
+							recordTracksEvent( 'calypso_jetpack_agency_dashboard_issue_license_button_click' )
+						)
+					}
+				>
+					{ translate( 'Issue License', { context: 'button label' } ) }
+				</Button>
+				<Button
+					primary
+					href="https://wordpress.com/jetpack/connect"
+					onClick={ () =>
+						dispatch(
+							recordTracksEvent( 'calypso_jetpack_agency_dashboard_add_site_button_click' )
+						)
+					}
+				>
+					{ translate( 'Add New Site', { context: 'button label' } ) }
+				</Button>
+			</div>
+		);
+	};
```
Non-blocking, but recommended: One thing I try to do in situations like this, is to create small additional components inside the same file, rather than build `renderThing` functions, just to keep the logic of the main component clear and simple. I realize the render function "pattern" is already present in this file, though, so I totally understand.

The alternative I'm proposing could look something like this:

```tsx
const AddSiteIssueLicenseButtons = () => {
	const dispatch = useDispatch();
	const translate = useTranslate();

	return (
			<div className="sites-overview__add-site-issue-license-buttons">
				<Button
					className="sites-overview__issue-license-button"
					href="/partner-portal/issue-license"
					onClick={ () =>
						dispatch(
							recordTracksEvent( 'calypso_jetpack_agency_dashboard_issue_license_button_click' )
						)
					}
				>
					{ translate( 'Issue License', { context: 'button label' } ) }
				</Button>
				<Button
					primary
					href="https://wordpress.com/jetpack/connect"
					onClick={ () =>
						dispatch(
							recordTracksEvent( 'calypso_jetpack_agency_dashboard_add_site_button_click' )
						)
					}
				>
					{ translate( 'Add New Site', { context: 'button label' } ) }
				</Button>
			</div>
		);
}
```

... and then the references would be replaced like so:

```tsx
<SiteContentHeader
	content={
		isLargeScreen &&
		( selectedLicensesCount > 0
			? renderIssueLicenseButton()
			: <AddSiteIssueLicenseButtons /> )
	}
```

```tsx
{ ! isLargeScreen && <AddSiteIssueLicenseButtons /> }
```

TheSteveK replied:
That's a good idea, and generally how I do that too, but I'm always a bit shy to do things completely differently when working with existing code like that.

mattgawarecki commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/style.scss:
```diff
@@ -35,13 +35,51 @@
 	}
 }
+.sites-overview__add-site-issue-license-buttons {
+	// We need these negative margin values because we want to make the container full-width,
+	// but our element is inside a limited-width parent.
```
Very nice leaving a comment like this, to explain the use of negative margins. (They always confuse me!)

TheSteveK replied:
Sadly, that one wasn't me, it was an existing comment I added to the new block since it behaved the same way, but thanks :)

mattgawarecki commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/style.scss:
```diff
@@ -35,13 +35,51 @@
+	}
+
+	.sites-overview__issue-license-button {
+		margin-inline-end: 0;
+		margin-bottom: 8px;
```
Non-blocking: I'd go with either one or the other notation style here -- either "inline-end" and "block-end," or "right" and "bottom."

mattgawarecki reviewed (commented):
I tested functionally and everything seems to work as described. I left a couple non-blocking suggestions, as well.

In my testing, I observed that if, e.g., a site already has Backup and I select Scan, the browser automatically navigates to the "Issue Licenses" page. Is this intentional? If so, consider this my approval. 👍

TheSteveK commented:
> In my testing, I observed that if, e.g., a site already has Backup and I select Scan, the browser automatically navigates to the "Issue Licenses" page. Is this intentional? If so, consider this my approval. 👍

I asked about this in a DM, but for transparency, that's the behavior of the existing add x licenses code, so it's not in scope for this PR

TheSteveK committed:
Refactor button render component

TheSteveK commented:
@mattgawarecki Alright, I checked all the boxes.  Have at :)

mattgawarecki commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/index.tsx:
```diff
@@ -242,13 +273,26 @@ export default function SitesOverview() {
 							pageTitle={ pageTitle }
-							showStickyContent={ !! showIssueLicenseButtonsLargeScreen }
+							// Only renderIssueLicenseButton should be sticky.
+							showStickyContent={ !! ( selectedLicensesCount > 0 && isLargeScreen ) }
```
Totally a nitpick: the `!!` here is redundant. It's a shortcut to convert any expression to a Boolean, but `selectedLicensesCount > 0 && isLargeScreen` is already a Boolean.

TheSteveK replied:
You would think, but typescript freaked out and the tests bombed at the build stage claiming the value could be null instead of a boolean

TheSteveK replied:
err undefined, not null :)

mattgawarecki replied:
Ahhh, I see. There's a fix for that too, but it's completely out of scope for this PR. 👍

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8632588

Thank you @TheSteveK for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/8632588).
